tests/scripts/analyze_outcomes.py

```diff
@@ -44,9 +47,40 @@ def hits(self):
         """
         return len(self.successes) + len(self.failures)
 
+class TestDescriptions(check_test_cases.TestDescriptionExplorer):
+    """Collect the available test cases."""
+
+    def __init__(self):
+        super().__init__()
+        self.descriptions = set()
+
+    def process_test_case(self, _per_file_state,
+                          file_name, _line_number, description):
+        """Record an available test case."""
+        base_name = re.sub(r'\.[^.]*$', '', re.sub(r'.*/', '', file_name))
+        key = ';'.join([base_name, description.decode('utf-8')])
+        self.descriptions.add(key)
+
+def collect_available_test_cases():
```

Comment:
Question: why use this function, rather than directly derive a CoverageAnalysis class from TestDescriptionExplorer, and define its `process_test_case()` method as the body of the `for` loop in `analyze_coverage()`?

Reply:
I don't understand the proposed alternative. The current structure of the code seems reasonable to me.

Reply:
I agree that the current structure is reasonable, this really was just a question, to know if you had considered and rejected the alternative for some reason.

The proposed alternative would be to not use this function, nor `class TestDescriptions`, and instead do something like:
```
class CoverageAnalyzer(check_test_cases.TestDescriptionExplorer):
    """Analyze coverage of the available test cases."""

    def __init__(self, outcomes, results):
        super().__init__()
        self.outcomes = outcomes
        self.results = results

    def process_test_case(self, _per_file_state,
                          file_name, _line_number, description):
        base_name = re.sub(r'\.[^.]*$', '', re.sub(r'.*/', '', file_name))
        key = ';'.join([base_name, description.decode('utf-8')])
        hits = outcomes[key].hits() if key in self.outcomes else 0
        if hits == 0:
            # Make this a warning, not an error, as long as we haven't
            # fixed this branch to have full coverage of test cases.
            self.results.warning('Test case not executed: {}', key)

def analyze_coverage(results, outcomes):
    """Check that all available test cases are executed at least once."""
    analyzer = CoverageAnalyzer(outcomes, results)
    analyzer.walk_all()
```
Now that I wrote it, I realise that your current code structure is probably more flexible if we want to add more analysis passes in the future, as with the alternative structure parts of the analyzer class would probably and up being duplicated.

Reply:
Oh, ok. I hadn't considered this approach. But looking at it, I prefer my way. There's a block of code that lists the available test cases, and when I'm doing the analysis, I don't care how it works. Your way entangles the way available test cases are determined with what we do with this information.